File: Application/Api/ProductHistories/List/Handlers/ListProductHistoriesHandler.cs
using Application.ErrorHandling.Application;
using Application.Interfaces.Persistence;
using MediatR;
using OneOf;

namespace Application.Api.ProductHistories.List.Handlers;

public class ListProductHistoriesHandler : IRequestHandler<ListProductHistoriesQuery, OneOf<ListProductHistoriesResult, List<PlainApplicationError>>>
{
    private readonly IProductHistoryRepository _productHistoryRepository;

    public ListProductHistoriesHandler(IProductHistoryRepository productHistoryRepository)
    {
        _productHistoryRepository = productHistoryRepository;
    }

    public async Task<OneOf<ListProductHistoriesResult, List<PlainApplicationError>>> Handle(ListProductHistoriesQuery request, CancellationToken cancellationToken)
    {
        var productHistories = await _productHistoryRepository.FindAllAsync(
            name: request.Name,
            minPrice: request.MinPrice,
            maxPrice: request.MaxPrice,
            description: request.Description,
            validFrom: request.ValidFrom,
            validTo: request.ValidTo,
            productId: request.ProductId
        );

        var result = new ListProductHistoriesResult(productHistories: productHistories);
        return result;
    }
}

File: frontend/src/application/commands/productHistories/list/ListProductHistoriesHandler.ts
import IProductHistoryDataAccess from "../../../interfaces/dataAccess/IProductHistoryDataAccess";
import { ICommandHandler } from "../../ICommandHandler";
import { err, ok } from "neverthrow";
import ListProductHistoriesCommand from "./ListProductHistoriesCommand";
import IListProductHistoriesResult from "./IListProductHistoriesResult";

export default class ListProductHistoriesHandler implements ICommandHandler<ListProductHistoriesCommand, IListProductHistoriesResult> {
    private _productHistoryDataAccess: IProductHistoryDataAccess;

    constructor(props: { productHistoryDataAccess: IProductHistoryDataAccess }) {
        this._productHistoryDataAccess = props.productHistoryDataAccess;
    }

    async handle(request: ListProductHistoriesCommand): Promise<IListProductHistoriesResult> {
        try {
            const result = await this._productHistoryDataAccess.listProductHistories({
                name: request.name,
                minPrice: request.minPrice,
                maxPrice: request.maxPrice,
                description: request.description,
                validTo: request.validTo,
                validFrom: request.validFrom,
                productId: request.productId,
                orderBy: request.orderBy
            });

            if (result.isErr()) {
                return err({ type: "API", data: result.error });
            }

            return ok({ productHistories: result.value });
        } catch (error: unknown) {
            return err({ type: "Exception", data: error });
        }
    }
}
